### PR TITLE
Migrating Vector Quantized Variational auto encoders to Keras3

### PR DESCRIPTION
**Pull request by LakshmiKalaKadali**
Major change: TensorFlow Probability is not backend-agnostic. To support JAX and PyTorch, I replaced tfp.layers.DistributionLambda with a custom sampling logic utilizing keras.random.categorical. This ensures the PixelCNN autoregressive generation works natively on all backends. Here is the reference 
[colab notebook](https://colab.sandbox.google.com/gist/LakshmiKalaKadali/7794143d0c7a494f6d0b3066cf9d93d5/vector_autoencoder.ipynb#scrollTo=JN_T0xcKj_4r).

### CODE REVIEW

File: examples/generative/vq_vae.py [hunk 1]
```diff
@@ -81,67 +90,57 @@ def __init__(self, num_embeddings, embedding_dim, beta=0.25, **kwargs):
         super().__init__(**kwargs)
         self.embedding_dim = embedding_dim
         self.num_embeddings = num_embeddings
-
-        # The `beta` parameter is best kept between [0.25, 2] as per the paper.
         self.beta = beta
```
Comment (sachinprasadhs) [inline examples/generative/vq_vae.py]: The comment was removed for this value, doesn't it hold good now?

File: examples/generative/vq_vae.py [hunk 2]
```diff
@@ -81,67 +90,57 @@ def __init__(self, num_embeddings, embedding_dim, beta=0.25, **kwargs):
...
 
-This line of code does the straight-through estimation part: `quantized = x +
-tf.stop_gradient(quantized - x)`. During backpropagation, `(quantized - x)` won't be
-included in the computation graph and the gradients obtained for `quantized`
-will be copied for `inputs`. Thanks to [this video](https://youtu.be/VZFVUrYcig0?t=1393)
+This line of code implements the straight-through estimator: quantized = x + ops.stop_gradient(quantized - x).
+In the forward pass, the terms cancel out (x+quantized−x), and the layer outputs the discrete quantized vectors.
+In the backward pass, since the gradient of ops.stop_gradient is zero,
+the gradient of the loss with respect to the output is effectively copied directly
+to the input x (the encoder's output).This allows the model to bypass the non-differentiable quantization step
```
Comment (sachinprasadhs) [inline examples/generative/vq_vae.py]: Add Space after (the encoder's output).

File: examples/generative/vq_vae.py [hunk 3]
```diff
@@ -81,67 +90,57 @@ def __init__(self, num_embeddings, embedding_dim, beta=0.25, **kwargs):
...
+and train the encoder using the decoder's gradients.. Thanks to [this video](https://youtu.be/VZFVUrYcig0?t=1393)
```
Comment (sachinprasadhs) [inline examples/generative/vq_vae.py]: Remove double period gradients.. --> gradients.

File: examples/generative/vq_vae.py [hunk 4]
```diff
@@ -230,45 +236,40 @@ def metrics(self):
...
+    def compute_loss(
+        self, x=None, y=None, y_pred=None, sample_weight=None, training=True
+    ):
+        # y_pred is now [reconstructions, indices]
+        reconstructions = y_pred[0]
 
-        # Backpropagation.
-        grads = tape.gradient(total_loss, self.vqvae.trainable_variables)
-        self.optimizer.apply_gradients(zip(grads, self.vqvae.trainable_variables))
+        # Rule 6: Stability
```
Comment (sachinprasadhs) [inline examples/generative/vq_vae.py]: What is this rule? Is it by mistake or are we refering to any predefined rules in the tutorial?

File: examples/generative/vq_vae.py [hunk 5]
```diff
@@ -230,45 +236,40 @@ def metrics(self):
...
 ## Load and preprocess the MNIST dataset
 """
 
 (x_train, _), (x_test, _) = keras.datasets.mnist.load_data()
+x_train = np.expand_dims(x_train, -1).astype("float32")
+x_test = np.expand_dims(x_test, -1).astype("float32")
 
-x_train = np.expand_dims(x_train, -1)
-x_test = np.expand_dims(x_test, -1)
+# Integrated Preprocessing (Rule 4)
```
Comment (sachinprasadhs) [inline examples/generative/vq_vae.py]: What is this rule #? Is it by mistake or are we refering to any predefined rules in the tutorial?

**Review by sachinprasadhs — COMMENTED** [review]
Just few doc comments, rest all looks good.
If regenerating tutorial takes time, you can just manually make the doc changes across all the files.

File: examples/generative/vq_vae.py (hunk 1 above)
Comment (LakshmiKalaKadali) [reply examples/generative/vq_vae.py]: Thanks for pointing this out. I added the comment back as beta has not changed.

File: examples/generative/vq_vae.py (hunk 2 above)
Comment (LakshmiKalaKadali) [reply examples/generative/vq_vae.py]: Done suggested changes

File: examples/generative/vq_vae.py (hunk 4 above)
Comment (LakshmiKalaKadali) [reply examples/generative/vq_vae.py]: By mistake. It was removed

File: examples/generative/vq_vae.py (hunk 5 above)
Comment (LakshmiKalaKadali) [reply examples/generative/vq_vae.py]: By mistake. It was removed

**Review by sachinprasadhs — APPROVED** [review]
LGTM, Thanks
